internal/provisioner/plain/controllers/bundleinstance_controller.go

```diff
@@ -262,12 +283,84 @@ func (r *BundleInstanceReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 	})
-	bi.Status.InstalledBundleName = bi.Spec.BundleName
+	bi.Status.InstalledBundleName = bundle.GetName()
+
+	if _, err := r.reconcileOldBundles(ctx, oldBundles, bundle); err != nil {
+		meta.SetStatusCondition(&bi.Status.Conditions, metav1.Condition{
+			Type:    rukpakv1alpha1.TypeProgressing,
+			Status:  metav1.ConditionFalse,
+			Reason:  rukpakv1alpha1.ReasonCleanupFailed,
+			Message: err.Error(),
+		})
+		return ctrl.Result{}, err
+	}
+	meta.RemoveStatusCondition(&bi.Status.Conditions, rukpakv1alpha1.TypeProgressing)
+
 	return ctrl.Result{}, nil
 }
 
+// reconcileDesiredBundle is responsible for checking whether the desired
+// Bundle resource that's specified in the BundleInstance parameter's
+// spec.Template configuration is present on cluster, and if not, creates
+// a new Bundle resource matching that desired specification.
+func (r *BundleInstanceReconciler) reconcileDesiredBundle(ctx context.Context, bi *rukpakv1alpha1.BundleInstance) (*rukpakv1alpha1.Bundle, []*rukpakv1alpha1.Bundle, error) {
+	// get the set of Bundle resources that already exist on cluster, and sort
+	// by metadata.CreationTimestamp in the case there's multiple Bundles
+	// that match the label selector.
+	existingBundles, err := util.GetBundlesForBundleInstanceSelector(ctx, r.Client, bi)
+	if err != nil {
+		return nil, nil, err
+	}
+	sort.Sort(util.BundlesByCreationTimestamp(existingBundles))
+
+	// check whether the BI controller has already reached the maximum
+	// generated Bundle limit to avoid hotlooping scenarios.
+	if len(existingBundles) > maxGeneratedBundleLimit {
+		return nil, nil, ErrMaxGeneratedLimit
+	}
+
+	// check whether there's an existing Bundle that matches the desired Bundle template
+	// specified in the BI resource, and if not, generate a new matches the template.
+	b := util.CheckExistingBundlesMatchesTemplate(existingBundles, bi.Spec.Template)
+	if b == nil {
+		controllerRef := metav1.NewControllerRef(bi, bi.GroupVersionKind())
+		b = &rukpakv1alpha1.Bundle{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            util.GenerateBundleName(bi.GetName()),
+				OwnerReferences: []metav1.OwnerReference{*controllerRef},
+				Labels:          bi.Spec.Template.Labels,
+			},
+			Spec: bi.Spec.Template.Spec,
+		}
+		if err := r.Create(ctx, b); err != nil {
+			return nil, nil, err
+		}
+	}
+	return b, existingBundles, err
+}
+
+// reconcileOldBundles is responsible for garbage collecting any Bundles
+// that no longer match the desired Bundle template.
+func (r *BundleInstanceReconciler) reconcileOldBundles(ctx context.Context, oldBundles []*rukpakv1alpha1.Bundle, currBundle *rukpakv1alpha1.Bundle) (int, error) {
+	var (
+		errors         []error
+		deletedBundles int
+	)
+	for _, b := range oldBundles {
+		if b.GetName() == currBundle.GetName() {
+			continue
+		}
+		if err := r.Delete(ctx, b); err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		deletedBundles++
+	}
+	return deletedBundles, utilerrors.NewAggregate(errors)
```

Comment:
Is there a point to return the `deletedBundles` count if we're not using it when we call this function above?